lib/sync/sync-server.js

```diff
@@ -0,0 +1,291 @@
+var ackProcessor = require('./ack-processor');
+var async = require('async');
+var dataHandlersModule = require('./dataHandlers');
+var datasets = require('./datasets');
+var defaultDataHandlersModule = require('./default-dataHandlers');
+var hashProviderModule = require('./hashProvider');
+var interceptorsModule = require('./interceptors');
+var metricsModule = require('./sync-metrics');
+var MongoClient = require('mongodb').MongoClient;
+var MongodbQueue = require('./mongodbQueue');
+var pendingProcessor = require('./pending-processor');
+var redis = require('redis');
+var storageModule = require('./storage');
+var syncApiModule = require('./api-sync');
+var syncLockModule = require('./lock');
+var syncProcessor = require('./sync-processor');
+var syncRecordsApiModule = require('./api-syncRecords');
+var syncSchedulerModule = require('./sync-scheduler');
+var syncUtil = require('./util');
+var util = require('util');
+var Worker = require('./worker');
+var _ = require('underscore');
+
+//TODO: remove redisClient. We probably don't need it anymore
+var redisClient = null;
+var mongoDbClient = null;
+var metricsClient = null;
+var hashProvider = null;
+var syncLock = null;
+var syncScheduler = null;
+var syncStorage = null;
+var interceptors = null;
+var apiSync = null;
+var apiSyncRecords = null;
+var dataHandlers = null;
+
+var ackQueue;
+var pendingQueue;
+var syncQueue;
+var ackWorker;
+var pendingWorker;
+var syncWorker;
+
+//default global configuration options for the sync server
+var DEFAULT_SYNC_CONF = {
+  //how often workers should check for the next job. Default: 1s
+  workerInterval: 1000,
+  //how often the scheduler should check the datasetClients. Default: 500ms
+  schedulerInterval: 500,
+  //the max time a scheduler can hold the lock for. Default: 20s
+  schedulerLockMaxTime: 20000,
+  //the default lock name for the sync scheduler
+  schedulerLockName: 'locks:sync:SyncScheduler'
+};
+var syncConfig = _.extend({}, DEFAULT_SYNC_CONF);
+var syncStarted = false;
+
+// Initialise cloud data sync service for specified dataset.
+function init(dataset_id, options, cb) {
+  syncUtil.doLog(dataset_id, 'info', 'init sync for dataset ' + dataset_id + ' with options ' + util.inspect(options));
+  datasets.init(dataset_id, options);
+  start(function(err) {
+    if (err) {
+      return cb(err);
+    }
+    syncStorage.updateManyDatasetClients({datasetId: dataset_id}, {stopped: false}, cb);
+  });
+}
+
+/**
+ * Connect to mongodb & redis with the given connection urls.
+ * This should be called before `start()` is called
+ *
+ * @param {string} mongoDBConnectionUrl A MongoDB connection uril in a format supported by the `mongodb` module
+ * @param {string} redisUrl A redis connection url in a format supported by the `redis` module
+ * @param {Object} metricsConf Metrics configuration
+ * @param {function} cb
+ */
+function connect(mongoDBConnectionUrl, redisUrl, metricsConf, cb) {
```

Comment:
I am wondering should we leave this one in the `index.js` file. This function is mainly to setup the prerequisites before we starting the sync server. Feels like it's not part of the sync server.